add plugin interface for collision detectors

mikeferguson opened the pull request:
This is a potential solution to #532. Highlights:
- Given the way the allocator stuff works (templated and CRTP), I chose not to wrap the CollisionRobot/World classes, but rather to create a new class that has an initialize() method in which one calls setActiveCollisionDetector() and does any other setup related to the collision detector.
- Parameter for the name of the collision plugin is held in "collision_detector", and I implemented a parameter search. In addition, if the parameter is not found, we check for "move_group/collision_detector", mainly so that the RVIZ plugin can use the collision detector.

@isucan please review and give feedback.

mikeferguson committed:
add plugin interface for collision detectors

davetcoleman commented on planning/collision_plugin/include/moveit/collision_plugin/collision_plugin.h:
```diff
@@ -0,0 +1,93 @@
+/*********************************************************************
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2014 Fetch Robotics Inc.
```
Super stealth mode...

davetcoleman commented on planning/collision_plugin/include/moveit/collision_plugin/collision_plugin.h:
```diff
@@ -0,0 +1,93 @@
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+#ifndef MOVEIT_COLLISION_PLUGIN_H
```
prefix with package name

mikeferguson replied:
whoops, looks like all the include guards are wrong -- i'll update those tomorrow.

davetcoleman commented:
+1 very clean code, i didn't compile though and don't have experience with collision checking

isucan commented on planning/collision_plugin/include/moveit/collision_plugin/collision_plugin.h:
```diff
@@ -0,0 +1,93 @@
+/*********************************************************************
```
This lib (collision_plugin) should be in moveit_core though, should it not?

mikeferguson replied:
This header could be in moveit_core -- I don't think you want the loader stuff in there though since it uses pluginlib. I'm fine with putting the header in moveit_core, as that is how the kinematics and planning_request_adapter stuff is done, just let me know and I can update the PR tomorrow morning.

isucan replied:
That is what I meant -- made a note about it below.

isucan commented on planning/collision_plugin/include/moveit/collision_plugin/collision_plugin_loader.h:
```diff
@@ -0,0 +1,73 @@
+/**
+ * @brief This is used to load the collision plugin
+ */
+class CollisionPluginLoader
```
This should be a separate library, indeed part of moveit_ros/planning

mikeferguson replied:
What do you mean by this? It is built as a separate library, but is not a separate package (as is done for the kinematics_plugin_loader).

isucan replied:
Sorry,  I did not look at the cmake file. My thought was only to have this
and the base class in different packages.
On Nov 30, 2014 10:56 PM, "Michael Ferguson" notifications@github.com
wrote:

> In
> planning/collision_plugin/include/moveit/collision_plugin/collision_plugin_loader.h:
> 
> > - \*  POSSIBILITY OF SUCH DAMAGE.
> > - ******************************************************************_**/
> >   +
> >   +#ifndef MOVEIT_COLLISION_PLUGIN_LOADER_H
> >   +#define MOVEIT_COLLISION_PLUGIN_LOADER_H
> >   +
> >   +#include <ros/ros.h>
> >   +#include <moveit/collision_plugin/collision_plugin.h>
> >   +
> >   +namespace collision_detection
> >   +{
> >   +
> >   +/_*
> > - \* @brief This is used to load the collision plugin
> > - */
> >   +class CollisionPluginLoader
> 
> What do you mean by this? It is built as a separate library, but is not a
> separate package (as is done for the kinematics_plugin_loader).
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ros-planning/moveit_ros/pull/536/files#r21072928.

isucan commented on planning/collision_plugin/src/collision_plugin_loader.cpp:
```diff
@@ -0,0 +1,150 @@
+
+private:
+  boost::shared_ptr<pluginlib::ClassLoader<CollisionPlugin> > loader_;
+  std::map<std::string, CollisionPlugin*> plugins_;
```
These seem to never be freed. We could have the map values be shared pointers / unique pointers maybe?

davetcoleman replied:
+1

isucan commented on planning/collision_plugin/src/collision_plugin_loader.cpp:
```diff
@@ -0,0 +1,150 @@
+  std::string param_name;
+  std::string collision_detector_name;
+
+  if (nh.searchParam("collision_detector", param_name))
```
Should these params be cached? If this function ends up being called often enough, it would be nicer to avoid calls to the param server.

isucan commented on planning/collision_plugin/src/collision_plugin_loader.cpp:
```diff
@@ -0,0 +1,150 @@
+  loader_->activate(name, scene, exclusive);
+}
+
+void CollisionPluginLoader::setupScene(
```
Return a bool indicating success?

isucan commented:
Very minor comments only, design is sound. This looks very close to be merged.

adolfo-rt commented on planning/collision_plugin/src/collision_plugin_loader.cpp:
```diff
@@ -0,0 +1,150 @@
+    bool exclusive)
+  {
+    std::map<std::string, CollisionPlugin*>::iterator it = plugins_.find(name);
+    if (it == plugins_.end())
```
Slightly OT, but I recently learned a [very compact](http://www.cplusplus.com/reference/map/map/count/) way to check whether a map contains a key-value pair. Performance-wise it's the same, but it's much less verbose.

``` c++
if (plugins_.count(name) == 0)
{
  // Key not found
}
else
{
  // Key found. Count == 1, as map is a unique associative container
}
```

davetcoleman replied:
cool!

mikeferguson commented:
Ok, I've moved collision_plugin.h to collision_detection folder of moveit_core. Also moved the collision_plugin folder to a better named collision_plugin_loader within this repo. Fixed potential memory leak by using boost::shared_ptr on the plugin instances.

I didn't cache the rosparams in this commit, since this is only called when a new instance of planningscenemonitor is created, however, that can certainly be a future improvement.

isucan commented:
Great! Thanks!
